app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/chat.tsx

```diff
@@ -0,0 +1,941 @@
+                        selectedReactions.has(emoji) ? '' : 'opacity-70 grayscale hover:grayscale-0'
+                    }`}
+                >
+                    <Twemoji key={emoji} emoji={emoji} size={REACTION_BUTTON_SIZE} />
+                </button>
+            ))}
+        </div>
+    )
+}
+
+// =================
+// === Reactions ===
+// =================lib/dashboard/src/authentication/src/dashboard/events/directoryListEvent.ts
+
+/** Props for a {@link Reactions}. */
+export interface ReactionsProps {
+    reactions: ReactionSymbol[]
+}
+
+/** A list of emoji reactions that have been on a message. */
+function Reactions(props: ReactionsProps) {
+    const { reactions } = props
+
+    if (reactions.length === 0) {
+        return null
+    } else {
+        return (
+            <div>
+                {reactions.map(reaction => (
+                    <Twemoji emoji={reaction} size={REACTION_SIZE} />
+                ))}
+            </div>
+        )
+    }
+}
+
+// ===================
+// === ChatMessage ===
+// ===================
+
+/** Props for a {@link ChatMessage}. */
+export interface ChatMessageProps {
+    message: ChatDisplayMessage
+    reactions: ReactionSymbol[]
+    shouldShowReactionBar: boolean
+    doReact: (reaction: ReactionSymbol) => void
+    doRemoveReaction: (reaction: ReactionSymbol) => void
+}
+
+/** A chat message, including user info, sent date, and reactions (if any). */
+function ChatMessage(props: ChatMessageProps) {
+    const { message, reactions, shouldShowReactionBar, doReact, doRemoveReaction } = props
+    return (
+        <div className="m-2">
+            <div className="flex">
+                <img
+                    crossOrigin="anonymous"
+                    src={message.avatar ?? DefaultUserIcon}
+                    className="rounded-full h-8 w-8 m-1"
+                />
+                <div className="m-1">
+                    <div className="font-semibold">{message.name}</div>
+                    <div className="text-opacity-50 text-primary">
+                        {dateTime.formatDateTimeChatFriendly(new Date(message.timestamp))}
+                    </div>
+                </div>
+            </div>
+            <div className="mx-1 whitespace-pre-wrap">
+                {message.content}
+                <Reactions reactions={reactions} />
+            </div>
+            {shouldShowReactionBar && (
+                <ReactionBar
+                    doReact={doReact}
+                    doRemoveReaction={doRemoveReaction}
+                    selectedReactions={new Set(message.reactions)}
+                />
+            )}
+        </div>
+    )
+}
+
+// ============
+// === Chat ===
+// ============
+
+/** Props for a {@link Chat}. */
+export interface ChatProps {
+    /** This should only be false when the panel is closing. */
+    isOpen: boolean
+    doClose: () => void
+}
+
+/** Chat sidebar. */
+function Chat(props: ChatProps) {
```

Comment:
this function is 400 lines long! Can we refactor it to several well named functions instead, please?